File: plusOne/src/plusOne/game.cpp
#include "game.h"
#include <iostream>
namespace plusOne
{
	std::mt19937 Game::randomSource;
	std::vector<std::pair<int, int> > Game::indices;
	std::vector<Game::searchState> Game::dfsStates;
	inline void intToDirection(int& x, int& y, int direction)
	{
		switch(direction)
		{
		case 0:
			x = 1; y = 0;
			break;
		case 1:
			x = 0; y = 1;
			break;
		case 2:
			x = -1, y = 0;
			break;
		case 3:
			x = 0; y = -1;
			break;
		}
	}
	Game::Game(int boardSize)
		: boardSize(boardSize), min(1), max(7), board(boardSize, boardSize)
	{
		std::uniform_int_distribution<> dist(min, max - 2);
		for(int i = 0; i < boardSize; i++)
		{
			for(int j = 0; j < boardSize; j++)
			{
				board(i, j) = dist(randomSource);
			}
		}
		std::uniform_int_distribution<> positionDist(0, boardSize - 1);
		int x = positionDist(randomSource);
		int y = positionDist(randomSource);
		board(y, x) = max - 1;
	}
	void Game::getIndicesForClick(int x, int y)
	{
		int currentValue = board(y, x);
		indices.clear();
		dfsStates.clear();

		mask = maskType::Constant(boardSize, boardSize, false);
		mask(y, x) = true;

		indices.push_back(std::make_pair(x, y));

		dfsStates.push_back(searchState(x, y, 0));
		while(dfsStates.size() > 0)
		{
			searchState& current = *dfsStates.rbegin();
			if(current.direction == 4)
			{
				dfsStates.pop_back();
			}
			else
			{
				int directionX, directionY;
				intToDirection(directionX, directionY, current.direction);
				current.direction++;
				int newX = current.x + directionX;
				int newY = current.y + directionY;
				if(newX < 0 || newY < 0 || newX >= boardSize || newY >= boardSize || mask(newY, newX) || board(newY, newX) != currentValue)
				{
					continue;
				}
				else
				{
					mask(newY, newX) = true;
					indices.push_back(std::make_pair(newX, newY));
					dfsStates.push_back(searchState(newX, newY, 0));
				}
			}
		}
	}
	void Game::incrementRange()
	{
		max++;
		min = std::max(1, max - 8);
	}
	void Game::shiftDown(Game::boardType& board, int boardSize)
	{
		for(int column = 0; column < boardSize; column++)
		{
			int row = boardSize - 1;
			bool allNegativeOne = true;
			for(int i = 0; i < row; i++) allNegativeOne &= board(i, column) == -1;
			while(row > 0 and !allNegativeOne)
			{
				if(board(row, column) == -1)
				{
					for(int i = row; i > 0; i--)
					{
						board(i, column) = board(i - 1, column);
					}
					board(0, column) = -1;
				}
				else row--;
				allNegativeOne = true;
				for(int i = 0; i < row; i++) allNegativeOne &= board(i, column) == -1;
			}
		}
	}
	void Game::fill(Game::boardType& board, int min, int max, int boardSize)
	{
		std::uniform_int_distribution<> dist(min, max - 2);
		for(int i = 0; i < boardSize; i++)
		{
			for(int j = 0; j < boardSize; j++)
			{
				if(board(i, j) == -1) board(i, j) = dist(randomSource);
			}
		}
	}
	std::pair<Game::boardType, int> Game::simulateClick(int x, int y)
	{
		if(x < 0 || y < 0 || x >= boardSize || y >= boardSize)
		{
			return std::make_pair(board, max);
		}
		int currentValue = board(y, x);
		Game::getIndicesForClick(x, y);
		if(indices.size() > 1)
		{
			boardType copiedBoard = board;
			for(std::pair<int, int> p : indices)
			{
				copiedBoard(p.second, p.first) = -1;
			}
			bool nextLevel;
			if(currentValue == max - 1)
			{
				nextLevel = true;
				int currentMin = copiedBoard.minCoeff();
				for(int i = 0; i < boardSize; i++)
				{
					for(int j = 0; j < boardSize; j++)
					{
						if(copiedBoard(i, j) == currentMin) copiedBoard(i, j) = -1;
					}
				}
			}
			else
			{
				nextLevel = false;
			}
			copiedBoard(y, x) = currentValue + 1;
			shiftDown(copiedBoard, boardSize);
			fill(copiedBoard, min + nextLevel, max + nextLevel, boardSize);
			if(nextLevel) return std::make_pair(copiedBoard, max + 1);
			return std::make_pair(copiedBoard, max);
		}
		return std::make_pair(board, max);
	}
	int Game::click(int x, int y)
	{
		if(x < 0 || y < 0 || x >= boardSize || y >= boardSize)
		{
			return -1;
		}
		int currentValue = board(y, x);
		Game::getIndicesForClick(x, y);
		if(indices.size() > 1)
		{
			for(std::pair<int, int> p : indices)
			{
				board(p.second, p.first) = -1;
			}
			bool nextLevel;
			if(currentValue == max - 1)
			{
				nextLevel = true;
				int currentMin = board.minCoeff();
				for(int i = 0; i < boardSize; i++)
				{
					for(int j = 0; j < boardSize; j++)
					{
						if(board(i, j) == currentMin) board(i, j) = -1;
					}
				}
				incrementRange();
			}
			else
			{
				nextLevel = false;
			}
			board(y, x) = currentValue + 1;
			shiftDown(board, boardSize);
			fill(board, min, max, boardSize);
			if(nextLevel) return 1;
			return 0;
		}
		return -1;
	}
	Game::boardType Game::getBoard()
	{
		return board;
	}
	bool Game::isValidMove(int x, int y)
	{
		int currentValue = board(y, x);
		for(int direction = 0; direction < 4; direction++)
		{
			int directionX, directionY;
			intToDirection(directionX, directionY, direction);
			int newX = x + directionX;
			int newY = y + directionY;
			if(newX < 0 || newY < 0 || newX >= boardSize || newY >= boardSize)
			{
				continue;
			}
			if(board(newY, newX) == currentValue) return true;
		}
		return false;
	}
	int Game::getMax()
	{
		return max;
	}
	void Game::setMax(int max)
	{
		max = max;
		min = std::max(1, max - 8);
	}
	void Game::setBoard(Game::boardType board)
	{
		this->board = board;
	}
	int Game::getBoardSize() const
	{
		return boardSize;
	}
}

File: plusOne/pysrc/plusOne/game.py
import numpy as np
class game:
    def __init__(self, boardSize):
        self._boardSize = boardSize
        self._board = np.random.randint(low = 1, high = 7, size = (boardSize, boardSize)) 
        self._min = 1
        self._max = 6
        self._mask = np.array([False] * boardSize * boardSize).reshape((boardSize, boardSize))

    def getIndicesForClick(self, x, y):
        currentValue = self._board[y, x]
        self._mask[:, :] = False
        directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        state = [(x, y, iter(directions))]
        self._mask[y, x] = True
        indices = [(x, y)]
        while len(state) > 0:
            x, y, direction = state[-1]
            nextDirection = next(direction, None)
            if nextDirection is None:
                state.pop()
            else:
                newX = x + nextDirection[0]
                newY = y + nextDirection[1]
                if newX < 0 or newX > self._boardSize - 1 or newY < 0 or newY > self._boardSize - 1 or self._mask[newY, newX] or self._board[newY, newX] != currentValue:
                    next
                else:
                    self._mask[newY, newX] = True
                    indices.append((newX, newY))
                    state.append((newX, newY, iter(directions)))
        return indices

    def click(self, x, y):
        x = int(x)
        y = int(y)
        if x < 0 or y < 0 or x >= self._boardSize or y >= self._boardSize:
            return(-1)
        else:
            currentValue = self._board[y, x]
            indices = self.getIndicesForClick(x, y)
            if len(indices) > 1:
                for (x, y) in indices:
                    self._board[y, x] = -1
                if currentValue == self._max:
                    nextLevel = True
                    currentMin = np.min(self._board)
                    for row in range(0, self._boardSize):
                        for column in range(0, self._boardSize):
                            if self._board[row, column] == currentMin:
                                self._board[row, column] = -1
                    self.increment_range()
                else:
                    nextLevel = False
                self._board[y, x] = currentValue + 1
                self.shift_down()
                self.fill()
                if nextLevel:
                    return(1)
                else:
                    return(0)
            return(-1)

    def increment_range(self):
        self._max = self._max + 1
        self._min = max(1, self._max - 8)

    def shift_down(self):
        for column in range(0, self._boardSize):
            row = self._boardSize - 1
            while row > 0 and ~np.all(self._board[:row, column] == -1):
                if self._board[row, column] == -1:
                    self._board[range(1, row+1), column] = self._board[range(0, row), column]
                    self._board[0, column] = -1
                else:
                    row = row - 1

    def fill(self):
        for column in range(0, self._boardSize):
            for row in range(0, self._boardSize):
                if self._board[row, column] == -1:
                    self._board[row, column] = np.random.randint(low = self._min, high = self._max)
    
    def isValidMove(self, x, y):
        if x < 0 or y < 0 or x >= self._boardSize or y >= self._boardSize:
            return(False)
        currentValue = self._board[y, x]
        directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        targets = [(direction[0] + y, direction[1] + x) for direction in directions]
        targets = [(y, x) for (y, x) in targets if not (x < 0 or y < 0 or x >= self._boardSize or y >= self._boardSize)]
        values = [self._board[y, x] for (y, x) in targets]
        return(currentValue in values)

    def isFailed(self):
        if np.all(self._board[1:, :] != self._board[:-1, :]) and np.all(self._board[:, 1:] != self._board[:, :-1]):
            return(True)
        return(False)
